### PR TITLE
Build should fail on a FAILED/SKIPPED test scenario 

### PR DESCRIPTION
**Pull request by jonnybbb**
Testcases to reproduce the issue of #52 and the actual fix for it.

### CODE REVIEW

File: plugin/src/main/java/org/gradle/testretry/internal/RetryTestResultProcessor.java [hunk 1]
```diff
@@ -68,11 +68,20 @@ public void started(TestDescriptorInternal descriptor, TestStartEvent testStartE
 
     @Override
     public void completed(Object testId, TestCompleteEvent testCompleteEvent) {
-        activeDescriptorsById.remove(testId);
+
+        final TestDescriptorInternal descriptor = activeDescriptorsById.get(testId);
```
Comment (marcphilipp) [inline plugin/src/main/java/org/gradle/testretry/internal/RetryTestResultProcessor.java:72]: nit: we don't seem to be consistent in our use of `final` in this class. We should either always use it when possible, or never.

Comment (marcphilipp) [reply plugin/src/main/java/org/gradle/testretry/internal/RetryTestResultProcessor.java:72]:
```suggestion
        final TestDescriptorInternal descriptor = activeDescriptorsById.remove(testId);
```

File: plugin/src/main/java/org/gradle/testretry/internal/RetryTestResultProcessor.java [hunk 2]
```diff
@@ -68,11 +68,20 @@ public void started(TestDescriptorInternal descriptor, TestStartEvent testStartE
...
+            final boolean hasFailedPreviously = nonExecutedFailedTests.remove(test);
+            if (hasFailedPreviously && testCompleteEvent.getResultType() == SKIPPED) {
+                failedTests.add(test);
+            }
+        }
 
         if (!testId.equals(rootTestDescriptorId) || lastRun()) {
             delegate.completed(testId, testCompleteEvent);
         }
+        activeDescriptorsById.remove(testId);
```
Comment (marcphilipp) [inline plugin/src/main/java/org/gradle/testretry/internal/RetryTestResultProcessor.java:84]:
```suggestion
```

File: plugin/src/main/java/org/gradle/testretry/internal/RetryTestResultProcessor.java [hunk 3]
```diff
@@ -84,7 +93,9 @@ public void output(Object testId, TestOutputEvent testOutputEvent) {
     public void failure(Object testId, Throwable throwable) {
         TestDescriptorInternal descriptor = activeDescriptorsById.get(testId);
         if (descriptor != null && descriptor.getClassName() != null) {
-            failedTests.add(new TestName(descriptor.getClassName(), descriptor.getName()));
+            final TestName test = new TestName(descriptor.getClassName(), descriptor.getName());
+            nonExecutedFailedTests.remove(test);
```
Comment (marcphilipp) [inline plugin/src/main/java/org/gradle/testretry/internal/RetryTestResultProcessor.java:97]: ❓ Why are we removing it here and in `completed`?

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy [hunk 4]
```diff
@@ -647,6 +648,130 @@ class SpockFuncTest extends AbstractPluginFuncTest {
...
+
+            @spock.lang.Stepwise
+            class StepwiseTests extends spock.lang.Specification {
+                def "parentTest"() {
+                    expect:
+                    true
+                }
+
+                @spock.lang.IgnoreIf({Files.exists(Paths.get("build/marker.file")) })
+                def "childTest"() {
```
Comment (marcphilipp) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:672]: ❓ Why "child" and "parent"?

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy [hunk 5]
```diff
@@ -647,6 +648,130 @@ class SpockFuncTest extends AbstractPluginFuncTest {
...
+        result.output.count('a flaky Test PASSED') == 0
+        result.output.contains('4 tests completed, 1 failed, 3 skipped')
+        !result.output.contains('Please file a bug report at')
+
+        where:
+        gradleVersion << GRADLE_VERSIONS_UNDER_TEST
+    }
+
+    @Unroll
+    def "build is successful if a test is ignored but never passed (gradle version #gradleVersion)"() {
```
Comment (marcphilipp) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:734]:
```suggestion
    def "build is successful if a test is ignored but never failed (gradle version #gradleVersion)"() {
```

File: plugin/src/main/java/org/gradle/testretry/internal/RetryTestResultProcessor.java (hunk 3 above)
Comment (jonnybbb) [reply plugin/src/main/java/org/gradle/testretry/internal/RetryTestResultProcessor.java:97]: previously it was executed beforehand in the `started` method which is called on every test execution. 
https://github.com/gradle/test-retry-gradle-plugin/pull/59/files#diff-c3de119e7a9886810b316ca4173f467eL55
But you are right its needless in the FAILED case,

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy (hunk 4 above)
Comment (jonnybbb) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:672]: The parent is indeed not required here anymore.

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy [hunk 6]
```diff
@@ -661,15 +661,10 @@ class SpockFuncTest extends AbstractPluginFuncTest {
             import java.nio.file.Paths
             import java.nio.file.Files
 
-            @spock.lang.Stepwise
             class StepwiseTests extends spock.lang.Specification {
```
Comment (marcphilipp) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:664]: This is no longer a "stepwise" test.

**Pull request merged**